### ISSUE TITLE
Idea for yubikey: Use U2F instead of HMAC

### ISSUE THREAD

**Issue by My1** [winauth/winauth#342]
This has mainly 2 reasons.
1) you dont need  a config slot on your yubi
2) it isnt limited to (in comparison, pretty expensive) yubikey but can be used with ANY U2F compatible device such as this 5€ stick for example: http://www.amazon.de/dp/B00OGPO3ZS/
these are from the user perspective, but from the security side you dont need any shared secrets for U2F since it's public key crypto.
in short it signs a challenge with its private key which can be verified using the public key.

both are challenge response so you cannot really encrypt but only authenticate the user, so there are no drawbacks.

but since there is no shared secret, no one can try to dive into the files and grab the secret to do the HMAC themselves.

**Comment by winauth** [comment]
U2F requires an authorisation server, WinAuth is intended to be standalone. I'm having trouble seeing how U2F is applicable. Interested hearing your solution.

**Comment by My1** [comment]
well winauth would hendle both the "server" and the "client" (usually browser) part by itself .
there are c libraries for both sides.

**Comment by winauth** [comment]
> winauth could handle both the "server" and the "client"

That's the point. If WinAuth is both, it isn't secure.

**Comment by My1** [comment]
but isnt this pretty much the same for HMAC?

**Comment by winauth** [comment]
No, because the the private key is stored in the Yubikey.

**Comment by My1** [comment]
and how to verify that? if you say public key now, it's literally the same for U2F

**Comment by My1** [comment]
let me quote this from the yubikey HMAC/challenge response page:
https://www.yubico.com/products/services-software/personalization-tools/challenge-response/
"YubiKey creates a “Response” based on a provided “Challenge” and a shared secret"

**Comment by My1** [comment]
The server only needs the pubkey of the U2F device (for verification), no shared secrets, no nothing.

**Comment by winauth** [comment]
The point is the key in HMAC is independent of WinAuth. That wouldn't be the case with U2F.

**Comment by winauth** [comment]
I just don't see how it could work. If WinAuth holds the key pair, it breaks. Can you come up with a sequence diagram?

**Comment by My1** [comment]
well winauth doesnt hold the key PAIR, just the pubkey+an identifier (plus attestation certs if you like).
https://developers.yubico.com/U2F/Protocol_details/Key_generation.html
this describes it somehow and I try to explain it with my words.

U2F works IIRC in a way like this:
you give the U2F stick an "AppID" (which essentially should be an HTTPS URL or a platform specific identifier like ios and android app-ID, let's just use localhost because it is local, I'll ask yubi support on how the ID should be for locally running applications) that is used to identify WHO asks for U2F

upon registration the U2F stick will throw you some things:
1) a signature make over the priv key
2) the pubkey
3) a "key Handle" (which is used by the device with a device-secret to create the key pair)
4) attestation certs (aka: this device is a real U2F device)

the certs prove that the key is a true U2F device , which may or may not be needed to be checked (the PHP library I use does that already iirc, so I dont actually care)

the signature with the pubkey gives you assurance that the device knows the related priv key.

and the key handle is the most important thing since this is what you need later (together with teh static appid) so that the U2F stick can re-create the keypair.

by the way how does HMAC even work if winauth cannot even know the secret (I mean it can use a pre-configged slot of a yubi)

**Comment by My1** [comment]
the only way I think HMAC would work without the secret would be to create 1 string let the yubi HMAC ir and store the result, but the kicks the challenge-response aspect into oblivion, and it can be killed by replacement attacks as well

**Comment by winauth** [comment]
I edited "key pair", meant challenger/challengee.

There are two things here. You can't separate the verification from the registration. Because WinAuth is open-source (or even if it were reverse-engineered), nothing can be secured without separate knowledge (e.g. password, Window account, yubikey). Registering with WinAuth and then verifying your U2F device is pointless. This is where I don't see it working without creating a separate online service, which I don't want to do.

But even if you can verify, you still can't protect the data. There is nothing that can be used to do this, so U2F doesn't add anything.

HMAC in a YubiKey works because the slot can be programmed independently. The encryption key is generated from a seed passed to the Yubikey. Obviously, it can still be read from memory, but that is a failing of a Windows based app not the protocol.

**Comment by My1** [comment]
wait okay that's not how you usually use a challenge response authorisation measure but it certainly is intresting: you have seed, let it HMAC and the result is used for encryption, that is actually a pretty intresting misuse of challenge-response but intresting, because on yubi pages it says that the whole HMAC part works on a shared secret, which is also the way I know it wo work, not for encryption but only verify.

(I just had a stupid idea on how it could, but actually couldnt work, as U2F has a counter)

**Comment by My1** [comment]
wait, I have an even weirder Idea. you could use the PUBLIC key of the U2F stick+handle combination for encryption with the handle being essentially the seed, and for actual control check you could verify the signature (which is only helpful for the UI but better than nothing) also since the pubkey is bound to the keyhandle, so without knowing that, well rest is obvious.
that would be essentially a similar approach to the HMAC misuse, where you use (according to how I catched it) the HMAC result for the encryption

**Comment by winauth** [comment]
Right, WinAuth is using HMAC for key generation, because (it being standalone) all it can do is encrypt data.

I wish they could make a more versatile and cheaper security key. Right now, the [YubiHSM](https://www.yubico.com/products/yubihsm/) would be perfect if everyone had one. It would provide absolute security because the encryption is done internally. Alas, at $500 a pop, it isn't exactly consumer-friendly.

**Comment by My1** [comment]
well yeah but by the way yubi has a smartcard function, which is essentially a cheap HSM

**Comment by My1** [comment]
and same as U2F, Smartcard doesnt need a config slot.

**Comment by winauth** [comment]
It's been a while since I did the Yubikey code, but as I remember setting up the PIV functionality required more effort on the user's part, whereas HMAC just worked. I would have to revisit it.

**Comment by My1** [comment]
also using the U2F public key is more or less the same as the HMAC misuse (especially because the U2F stick uses HMAC for throwing the keyhandle and the device secret together

**Comment by My1** [comment]
also in that case you would only need the client library.

since I misinterpreted the HMAC in the opening post (since I thought it's just verification)
the advantages and disadvantages shift a bit, but it still remains that U2F has larger HW support (5€ keys in comparison to 40-60€ for a yubi, depending on the model and store)

**Comment by My1** [comment]
the only disadvantage (or advantage, depending on how you see it) is that U2F cannot be backed up because the device secret is already burnt in when they are made, so it would be intresting to create a functionality to server multiple different U2Fs.

**Comment by winauth** [comment]
> also using the U2F public key is more or less the same as the HMAC misuse (especially because the U2F stick uses HMAC for throwing the keyhandle and the device secret together

That may well be possible. And as long as the libraries work with all devices.

**Comment by My1** [comment]
well U2F isnt a standard without reason and it's an HID so it should work, I dont know about how the c library works but at least my browser works with both the yubi 4 nano and the mega cheap U2F stick linked earlier.

**Comment by winauth** [comment]
Backups should be user dependant and not per device. No different than forgetting your password. You're screwed unless you have plain text backups.

**Comment by My1** [comment]
yep that's also a way to see it.

**Comment by My1** [comment]
also one of the most epic thing is, if you have a NFC compatible U2F device you could even make hardware auth possible on an android app, when you want to plan that, which I would really like because many apps dont have properly secured internal storage (I mean look at steam) for the 2FA data.

**Comment by My1** [comment]
doe this mean this wont be done? too sad.

**Comment by winauth** [comment]
I'm taking it out of the active bug list (cleaning up) and marked as enhancement. I have a note of this along with smartcard features., but whether I get to it is another thing.
